File: src/test/java/com/thebluehats/server/game/enchants/ComboSwiftTest.java
package com.thebluehats.server.game.enchants;

import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.spy;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

import com.thebluehats.server.game.enchants.args.common.PotionEffectWithHitsNeededArgs;
import com.thebluehats.server.game.managers.combat.templates.EventTemplate;
import com.thebluehats.server.game.managers.combat.templates.PlayerHitPlayerTemplate;
import com.thebluehats.server.game.managers.enchants.CustomEnchant;
import com.thebluehats.server.game.managers.enchants.HitCounter;

import org.bukkit.entity.Player;
import org.bukkit.potion.PotionEffect;
import org.bukkit.potion.PotionEffectType;
import org.junit.Test;

public class ComboSwiftTest {

    @Test
    public void SpeedGivenIfHitsNeededAreReached() {
        final int HITS_NEEDED = 5;
        final int SPEED_TIME = 2;
        final int AMPLIFIER = 1;

        Player player = mock(Player.class);
        HitCounter hitCounter = mock(HitCounter.class);

        CustomEnchant<PotionEffectWithHitsNeededArgs> comboSwift = new ComboSwift(hitCounter,
                new EventTemplate[] { new PlayerHitPlayerTemplate() });

        PotionEffectWithHitsNeededArgs args = spy(
                new PotionEffectWithHitsNeededArgs(player, SPEED_TIME, AMPLIFIER, HITS_NEEDED));

        when(hitCounter.hasHits(player, HITS_NEEDED)).thenReturn(true);
        when(args.getPlayer()).thenReturn(player);

        comboSwift.execute(args);
        verify(player).addPotionEffect(new PotionEffect(PotionEffectType.SPEED, SPEED_TIME * 20, AMPLIFIER, true));
    }

    @Test
    public void SpeedNotGivenIfHitsNeededAreNotReached() {
        final int HITS_NEEDED = 5;
        final int SPEED_TIME = 2;
        final int AMPLIFIER = 1;

        Player player = mock(Player.class);
        HitCounter hitCounter = mock(HitCounter.class);

        ComboSwift comboSwift = new ComboSwift(hitCounter, new EventTemplate[] { new PlayerHitPlayerTemplate() });

        PotionEffectWithHitsNeededArgs args = spy(
                new PotionEffectWithHitsNeededArgs(player, SPEED_TIME, AMPLIFIER, HITS_NEEDED));

        when(hitCounter.hasHits(player, HITS_NEEDED)).thenReturn(true);
        when(args.getPlayer()).thenReturn(player);

        comboSwift.execute(args);
        verify(player).addPotionEffect(new PotionEffect(PotionEffectType.SPEED, SPEED_TIME * 20, AMPLIFIER, true));
    }
}

File: src/test/kotlin/com/thebluehats/server/game/enchants/ComboSwiftTest.kt
package com.thebluehats.server.game.enchants

import com.thebluehats.server.game.managers.enchants.HitCounter
import com.thebluehats.server.game.managers.enchants.processedevents.DamageEventEnchantData
import io.kotest.core.spec.style.DescribeSpec
import io.mockk.every
import io.mockk.mockk
import io.mockk.verify
import org.bukkit.entity.Player
import org.bukkit.potion.PotionEffect
import org.bukkit.potion.PotionEffectType
import java.util.*

class ComboSwiftTest : DescribeSpec ({
    describe("combo swift") {
        it("should give speed to the damager") {
            val effect = PotionEffect(PotionEffectType.SPEED, 5 * 20, 1)

            val damagerMock = mockk<Player> {
                every { addPotionEffect(effect,true) } returns true
            }

            val hitCounterMock = mockk<HitCounter> {
                every { addOne(damagerMock) } returns Unit
                every { hasHits(damagerMock, 3) } returns true
            }

            val data = mockk<DamageEventEnchantData> {
                every { level } returns 3
                every { damager} returns damagerMock
            }

            val comboSwift = ComboSwift(hitCounterMock, mockk())

            comboSwift.execute(data)

            verify { damagerMock.addPotionEffect(effect, true) }
        }
    }
})